runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,757 @@
+	session_info::{AccountId, IdentificationTuple},
+};
+use frame_support::{
+	traits::{Defensive, Get, KeyOwnerProofSystem, ValidatorSet, ValidatorSetWithIdentification},
+	weights::{Pays, Weight},
+};
+use parity_scale_codec::{Decode, Encode};
+use primitives::v2::{CandidateHash, SessionIndex, ValidatorId, ValidatorIndex};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::Convert,
+	transaction_validity::{
+		InvalidTransaction, TransactionPriority, TransactionSource, TransactionValidity,
+		TransactionValidityError, ValidTransaction,
+	},
+	DispatchResult, KeyTypeId, Perbill, RuntimeDebug,
+};
+use sp_session::{GetSessionNumber, GetValidatorCount};
+use sp_staking::offence::{DisableStrategy, Kind, Offence, OffenceError, ReportOffence};
+use sp_std::{
+	collections::btree_map::{BTreeMap, Entry},
+	prelude::*,
+};
+
+const LOG_TARGET: &str = "runtime::parachains::slashing";
+
+// These are constants, but we want to make them configurable
+// via `HostConfiguration` in the future.
+const SLASH_FOR_INVALID: Perbill = Perbill::from_percent(100);
+const SLASH_AGAINST_VALID: Perbill = Perbill::from_percent(1);
+const DEFENSIVE_PROOF: &'static str = "disputes module should bail on old session";
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking;
+
+/// The benchmarking configuration.
+pub trait BenchmarkingConfiguration {
+	const MAX_VALIDATORS: u32;
+}
+
+pub struct BenchConfig<const M: u32>;
+
+impl<const M: u32> BenchmarkingConfiguration for BenchConfig<M> {
+	const MAX_VALIDATORS: u32 = M;
+}
+
+/// Timeslots should uniquely identify offences and are used for the offence
+/// deduplication.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+pub struct DisputesTimeSlot {
+	// The order of these matters for `derive(Ord)`.
+	session_index: SessionIndex,
+	candidate_hash: CandidateHash,
+}
+
+impl DisputesTimeSlot {
+	pub fn new(session_index: SessionIndex, candidate_hash: CandidateHash) -> Self {
+		Self { session_index, candidate_hash }
+	}
+}
+
+/// An offence that is filed when a series of validators lost a dispute.
+#[derive(RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct SlashingOffence<KeyOwnerIdentification> {
+	/// The size of the validator set in that session.
+	pub validator_set_count: ValidatorSetCount,
+	/// Should be unique per dispute.
+	pub time_slot: DisputesTimeSlot,
+	/// Staking information about the validators that lost the dispute
+	/// needed for slashing.
+	pub offenders: Vec<KeyOwnerIdentification>,
+	/// What fraction of the total exposure that should be slashed for
+	/// this offence.
+	pub slash_fraction: Perbill,
+	/// Whether the candidate was valid or invalid.
+	pub kind: SlashingOffenceKind,
+}
+
+impl<Offender> Offence<Offender> for SlashingOffence<Offender>
+where
+	Offender: Clone,
+{
+	const ID: Kind = *b"disputes:slashin";
+
+	type TimeSlot = DisputesTimeSlot;
+
+	fn offenders(&self) -> Vec<Offender> {
```

Comment:
This is an implementation of a trait from substrate. Not sure what you're suggesting. Changing the type of `Offender` to be a reference? Not sure if this will work.